### PR TITLE
Add CondVar redux

### PR DESCRIPTION
**Pull request by dyfer**
Cleaned up PR #5394

Description from the commit message:

A condition variable implementation that intends to replace Condition.
Main features:
- a more intuitive interface
- stronger guarantees about exception safety and blocking/unblocking
  conditions
- predicates are given per-wait instead of per-object
- timeouts
- more closely matches classic condition variable design and what's
  found in other implementations

## Types of changes

- New feature

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by patrickdupuis** [comment]
@dyfer are the two failing tests expected?

**Comment by dyfer** [comment]
There's a [CondVar test that fails](https://github.com/supercollider/supercollider/pull/5436/checks?check_run_id=2487618726#step:9:13623) - I wouldn't expect that, not sure why this wasn't the case in the other PR

**Comment by patrickdupuis** [comment]
Look like it failed on the second assert inside `TestCondVar.test_waitFor_withPredicate_signaledButFalseSoTimedOut`
```
PASS: a TestCondVar: test_waitFor_withPredicate_signaledButFalseSoTimedOut - waitResult
Is:
	 false 
Should be:
	 false

FAIL: a TestCondVar: test_waitFor_withPredicate_signaledButFalseSoTimedOut - counter
Is:
	 2 
Should be:
	 3
```

File: testsuite/classlibrary/TestCondVar.sc [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+		// This wait has to be a bit longer because we need to ensure there's enough time for the spurious wakeup before
+		// timing out.
+		waitResult = c.waitFor(0.1) {
+			counter = counter + 1;
+			false
+		};
+		this.assertEquals(waitResult, false, "waitResult");
+		// Evaluated 3 times -- when blocked, when signalled, when timed out
+		this.assertEquals(counter, 3, "counter");
```
Comment (patrickdupuis) [inline testsuite/classlibrary/TestCondVar.sc:307]: As the comment indicates, this test assumes the CondVar blocked *before* line 297 signaled it. There's no guarantee things will happen in that order. The fork could signal before `CondVar.waitFor` resulting in an incorrect `counter`. On my machine, this test passes. On GA, it appears to fail at least some of the time. We should make this test behave in a predictable manner.

File: testsuite/classlibrary/TestCondVar.sc [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+	test_waitFor_withPredicate_signaledButFalseSoTimedOut {
+		var c = CondVar();
+		var counter = 0;
+		var waitResult;
+		fork { c.signalOne };
+
+		// This wait has to be a bit longer because we need to ensure there's enough time for the spurious wakeup before
+		// timing out.
+		waitResult = c.waitFor(0.1) {
```
Comment (patrickdupuis) [inline testsuite/classlibrary/TestCondVar.sc:301]: I've never seen this syntax before. I can understand writing `c.wait { /* */ }` but not so much `c.waitFor(1) { /* */ }`. Maybe I'm just not used to it. What do others think? @joshpar @telephon @jamshark70

**Review by patrickdupuis — CHANGES_REQUESTED** [review]
We need to make a few changes to these tests. I'm going to try and see what I can come up with.

**Comment by patrickdupuis** [comment]
@dyfer this test passed in #5394 due to the printing that we removed in the PR:
https://github.com/supercollider/supercollider/pull/5394/commits/4100269b1c4d22ca6e939851c33d85b7764f0206

File: testsuite/classlibrary/TestCondVar.sc [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+	// wait - without predicate
+	test_wait_withoutPredicate_oneThread_signalOne {
+		var c = CondVar();
+		var waitResult, signalOneResult;
+		fork {
+			signalOneResult = c.signalOne;
+		};
+
+		// will timeout if signalOne does not wake us
+		waitResult = c.wait;
```
Comment (patrickdupuis) [inline testsuite/classlibrary/TestCondVar.sc:29]: Nothing will cause this test to timeout. The test will hang indefinitely if `c.signalOne` happens just before `c.wait`. One way to fix this would be to schedule `c.signalOne` some small amount of time into the future.

Comment (patrickdupuis) [reply testsuite/classlibrary/TestCondVar.sc:29]: This isn't something I'm very knowledgeable in. Would adding this to the fork be an improvement?
```
fork {
    thisThread.clock.sched(this.smallTimeSlice, { signalOneResult = c.signalOne });
};
```

**Comment by patrickdupuis** [comment]
I restarted the CI and the same failures happened. They don't seem to be random.

**Comment by dyfer** [comment]
> @dyfer this test passed in #5394 due to the printing that we removed in this PR:
> [4100269](https://github.com/supercollider/supercollider/commit/4100269b1c4d22ca6e939851c33d85b7764f0206)

Ah I see. I missed that there's a difference in timing for `waitFor` in that commit. I'll try to change that value and see if it makes the test pass...

**Comment by dyfer** [comment]
With the increased timing the test passes. Is this good to go then? what do you think @patrickdupuis ?

File: testsuite/classlibrary/TestCondVar.sc (hunk 2 above)
Comment (jamshark70) [reply testsuite/classlibrary/TestCondVar.sc]: In a way, we've all seen this syntax before, with a slight difference:

- `if(receiver) { block } { block }`
- `switch(receiver) { block } { block } { block }...`

The difference here is: `receiver.waitFor(arg0) { block }`.

My feeling is that this is not a massive difference, and it doesn't bother me. However, Brian balked at it in the earlier PR (so I'm surprised to see it appear in the test suite).

This is one where, no matter what you do, somebody can complain about something:

- If we recommend `c.wait { xyz }` and `c.waitFor(time, { xyz })`, then the bracketing is inconsistent between the two forms.

- If we recommend `c.wait({ xyz })` and `c.waitFor(time, { xyz })`, someone will be bothered by the double brackets for `wait`.

- If we recommend `c.wait { xyz }` and `c.waitFor(time) { xyz })`... well, the present question is exactly a complaint about this.

My personal opinion is to favor trailing-function syntax in general, especially for methods related to flow of control. (`wait` is not a branching keyword, but CondVar is a control-flow device.) If we are comfortable with `if(x) { y }` then I see no reason to be uncomfortable with `cond.waitFor(num) { y }`. For that matter, I never got in the habit of writing e.g. `do(10) { body }` but I wouldn't object to it (and in some ways I think it reads even better than `10.do`).

TL;DR If it's up to me, I wouldn't change it, but I won't object if it really troubles others.

File: testsuite/classlibrary/TestCondVar.sc (hunk 3 above)
Comment (patrickdupuis) [reply testsuite/classlibrary/TestCondVar.sc:29]: I don't want to keep this PR from being merged. I'm resolving this conversation.

File: testsuite/classlibrary/TestCondVar.sc (hunk 1 above)
Comment (patrickdupuis) [reply testsuite/classlibrary/TestCondVar.sc:307]: I'm resolving this conversation.

File: testsuite/classlibrary/TestCondVar.sc (hunk 2 above)
Comment (patrickdupuis) [reply testsuite/classlibrary/TestCondVar.sc]: Thanks for the explanation! I agree, we should just keep it as is.

**Comment by patrickdupuis** [comment]
@dyfer Yes, we are good to go!

**Review by patrickdupuis — APPROVED** [review]
